cleanrl/cql_continuous/sac_cql.py

```diff
@@ -0,0 +1,396 @@
+        self.dataset = self.dataset_env.get_dataset()
+    def __len__(self):
+        return self.dataset['observations'].shape[0]-1
+    def __getitem__(self, index):
+        return self.dataset['observations'][:-1][index].astype(np.float32), \
+               self.dataset['actions'][:-1][index].astype(np.float32), \
+               self.dataset['rewards'][:-1][index].astype(np.float32), \
+               self.dataset['observations'][1:][index].astype(np.float32), \
+               self.dataset['terminals'][:-1][index].astype(np.float32)
+
+data_loader = DataLoader(ExperienceReplayDataset(args.offline_gym_id), batch_size=args.batch_size, num_workers=2)
+
+pg = Policy(input_shape, output_shape, env).to(device)
+qf1 = SoftQNetwork(input_shape, output_shape, layer_init).to(device)
+qf2 = SoftQNetwork(input_shape, output_shape, layer_init).to(device)
+qf1_target = SoftQNetwork(input_shape, output_shape, layer_init).to(device)
+qf2_target = SoftQNetwork(input_shape, output_shape, layer_init).to(device)
+qf1_target.load_state_dict(qf1.state_dict())
+qf2_target.load_state_dict(qf2.state_dict())
+values_optimizer = optim.Adam(list(qf1.parameters()) + list(qf2.parameters()), lr=args.q_lr)
+policy_optimizer = optim.Adam(list(pg.parameters()), lr=args.policy_lr)
+loss_fn = nn.MSELoss()
+
+# Automatic entropy tuning for SAC
+if args.autotune:
+    target_entropy = - torch.prod(torch.Tensor(env.action_space.shape).to(device)).item()
+    log_alpha = torch.zeros(1, requires_grad=True, device=device)
+    alpha = log_alpha.exp().item()
+    a_optimizer = optim.Adam([log_alpha], lr=args.q_lr)
+else:
+    alpha = args.alpha
+
+# Automatic tuming for the min Q term (alpha_prime) coefficient
+if args.with_lagrange:
+    log_alpha_prime = torch.zeros([1], requires_grad=True, device=device)
+    alpha_prime_optimizer = optim.Adam([log_alpha_prime], args.alpha_lr)
+
+# Helper to evaulate the agent, could use some clean up.
+def test_agent(env, policy, n_eval_episodes=5):
+    returns, lengths = [], []
+
+    for _ in range(n_eval_episodes):
+        ret, done, t = 0., False, 0
+        obs = np.array( env.reset())
+
+        while not done:
+            # MaxEntRL Paper argues eval should not be determinsitic
+            with torch.no_grad():
+                action, _, _ = policy.get_action(torch.Tensor(obs).unsqueeze(0).to(device))
+                action = action.tolist()[0]
+
+            obs, rew, done, _ = env.step( action)
+            obs = np.array( obs)
+            ret += rew
+            t += 1
+
+        returns.append(ret)
+        lengths.append(t)
+
+    eval_stats = {
+        "test_mean_return": np.mean(returns),
+        "test_mean_length": np.mean(lengths),
+        "test_max_return": np.max(returns),
+        "test_max_length": np.max(lengths),
+        "test_min_return": np.min(returns),
+        "test_min_length": np.min(lengths)
+    }
+
+    return eval_stats
+
+# Training Loop
+global_step = 0 # Tracks the Gradient updates
```

Comment:
Maybe setup the loop in the following way instead:

```
for global_step in range(args.total_timesteps):
    if global_step % args.train_frequency == 0:
        s_obs, s_actions, s_rewards, s_next_obses, s_dones = next(data_loader)
        s_obs, s_actions, s_rewards, s_next_obses, s_dones = s_obs.to(device), s_actions.to(device), s_rewards.to(device), s_next_obses.to(device), s_dones.to(device)
        # ..... training code
    if global_step % args.eval_frequency == 0:
        #  .....  eval code
        # 
```

This will have similar structure as the CQL code.

Reply:
- Removed the `n-epochs`  and `epoch-length` parser argument, replaced by `total-timesteps`.
- Modified the dataset class to match that in `offline_dqn_cql` (IterableDataset, etc...)
- Update the training loop to match the one in your comment, to a certain extent:

```
for global_step in range(args.total_timesteps):
    # Oneliner batch sample and moving to device ?
    s_obs, s_actions, s_rewards, s_next_obses, s_dones = [i.to(device) for i in next(data_loader)]

    # ..... training code
    
    if global_step % args.eval_frequency == 0:
        #  .....  eval code
        #  .... logging various metrics to wandb, etc...
```